# Importing Libraries 

In [2]:
import spacy
from spacy.lang.en.stop_words import STOP_WORDS
from string import punctuation

# List of Stop Words

In [3]:
stopwords = list(STOP_WORDS)

# Text

In [4]:
text = """"Mahmoud Darwish is a Palestinian poet and member of the Palestinian National Council of the Palestine Liberation Committee. He has poetry collections full of modernist content. He was born in 1941 in the village of Al-Birwa, a Palestinian village located in the Galilee near the coast of Acre, where his family owned land. The family went out with the Palestinian refugees in 1948 to Lebanon, then returned infiltrated in 1949 after the armistice was signed, only to find the village destroyed and a moshav (an Israeli agricultural village) “Ahihud” built on its lands. Kibbutz Yas'ur lived with his family in the new village.
After sending his secondary education in a high school in high school, he joined the Israeli Communist Party, the new party in the party, and the new one who edited it, edited it, edited it, edited Al-Fajr newspaper, which was published by Mapam.
He served as president of the General Union of Palestinian Writers and Writers and edited Al-Karmel magazine. His residence was in Paris before his return to his homeland, as he entered Palestine with a permit to visit his mother. While he was there, some Arab and Jewish members of the Israeli Knesset made a proposal to allow him to stay, and he was allowed to do so.
Critics divided the stages of Darwish's poetry into several sections that combine the poet's relationship with his homeland, his cause "the Palestinian cause", exile and leaving home, all in light of his relationship with the self. Critic Muhammad Fikri Al-Jazzar divided Darwish's poetry into three sections: The first stage, which is the stage of his presence in the homeland, which includes the beginnings of the poet's formation and awareness of the cause of his homeland and the formation of belonging to this homeland under the occupation."""

In [7]:
# Object to process the text 
nlp = spacy.load('en_core_web_sm')

In [8]:
# Segminting the text into words, punctuation and so on
doc = nlp(text)

In [9]:
#each doc consists of individual tokens
# List of tokens
tokens = [token.text for token in doc]
print(tokens)

['"', 'Mahmoud', 'Darwish', 'is', 'a', 'Palestinian', 'poet', 'and', 'member', 'of', 'the', 'Palestinian', 'National', 'Council', 'of', 'the', 'Palestine', 'Liberation', 'Committee', '.', 'He', 'has', 'poetry', 'collections', 'full', 'of', 'modernist', 'content', '.', 'He', 'was', 'born', 'in', '1941', 'in', 'the', 'village', 'of', 'Al', '-', 'Birwa', ',', 'a', 'Palestinian', 'village', 'located', 'in', 'the', 'Galilee', 'near', 'the', 'coast', 'of', 'Acre', ',', 'where', 'his', 'family', 'owned', 'land', '.', 'The', 'family', 'went', 'out', 'with', 'the', 'Palestinian', 'refugees', 'in', '1948', 'to', 'Lebanon', ',', 'then', 'returned', 'infiltrated', 'in', '1949', 'after', 'the', 'armistice', 'was', 'signed', ',', 'only', 'to', 'find', 'the', 'village', 'destroyed', 'and', 'a', 'moshav', '(', 'an', 'Israeli', 'agricultural', 'village', ')', '“', 'Ahihud', '”', 'built', 'on', 'its', 'lands', '.', 'Kibbutz', "Yas'ur", 'lived', 'with', 'his', 'family', 'in', 'the', 'new', 'village', '.'

In [10]:
# Removing punctuation & Adding '\n'
punctuation = punctuation + '\n'
punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~\n'

In [11]:
# Counting the number of occurrences of all the distinct words which are not punctuations or stopwords 
word_frequencies = {}
for word in doc:
    if word.text.lower() not in stopwords:
        if word.text.lower() not in punctuation:
            if word.text not in word_frequencies.keys():
                word_frequencies[word.text] = 1
            else:
                word_frequencies[word.text] += 1
                
print(word_frequencies)

{'Mahmoud': 1, 'Darwish': 3, 'Palestinian': 6, 'poet': 3, 'member': 1, 'National': 1, 'Council': 1, 'Palestine': 2, 'Liberation': 1, 'Committee': 1, 'poetry': 3, 'collections': 1, 'modernist': 1, 'content': 1, 'born': 1, '1941': 1, 'village': 5, 'Al': 4, 'Birwa': 1, 'located': 1, 'Galilee': 1, 'near': 1, 'coast': 1, 'Acre': 1, 'family': 3, 'owned': 1, 'land': 1, 'went': 1, 'refugees': 1, '1948': 1, 'Lebanon': 1, 'returned': 1, 'infiltrated': 1, '1949': 1, 'armistice': 1, 'signed': 1, 'find': 1, 'destroyed': 1, 'moshav': 1, 'Israeli': 3, 'agricultural': 1, '“': 1, 'Ahihud': 1, '”': 1, 'built': 1, 'lands': 1, 'Kibbutz': 1, "Yas'ur": 1, 'lived': 1, 'new': 3, 'sending': 1, 'secondary': 1, 'education': 1, 'high': 2, 'school': 2, 'joined': 1, 'Communist': 1, 'Party': 1, 'party': 2, 'edited': 5, 'Fajr': 1, 'newspaper': 1, 'published': 1, 'Mapam': 1, 'served': 1, 'president': 1, 'General': 1, 'Union': 1, 'Writers': 2, 'Karmel': 1, 'magazine': 1, 'residence': 1, 'Paris': 1, 'return': 1, 'homela

In [12]:
# Get the max_frequency: the most repeated word
max_frequency = max(word_frequencies.values())
max_frequency

6

# Normalizing the Frequencies

In [13]:
# We will divide each frequency value in word_frequencies with the max_frequency to normalize the frequencies.
for word in word_frequencies.keys():
    word_frequencies[word] = word_frequencies[word] / max_frequency

print(word_frequencies)

{'Mahmoud': 0.16666666666666666, 'Darwish': 0.5, 'Palestinian': 1.0, 'poet': 0.5, 'member': 0.16666666666666666, 'National': 0.16666666666666666, 'Council': 0.16666666666666666, 'Palestine': 0.3333333333333333, 'Liberation': 0.16666666666666666, 'Committee': 0.16666666666666666, 'poetry': 0.5, 'collections': 0.16666666666666666, 'modernist': 0.16666666666666666, 'content': 0.16666666666666666, 'born': 0.16666666666666666, '1941': 0.16666666666666666, 'village': 0.8333333333333334, 'Al': 0.6666666666666666, 'Birwa': 0.16666666666666666, 'located': 0.16666666666666666, 'Galilee': 0.16666666666666666, 'near': 0.16666666666666666, 'coast': 0.16666666666666666, 'Acre': 0.16666666666666666, 'family': 0.5, 'owned': 0.16666666666666666, 'land': 0.16666666666666666, 'went': 0.16666666666666666, 'refugees': 0.16666666666666666, '1948': 0.16666666666666666, 'Lebanon': 0.16666666666666666, 'returned': 0.16666666666666666, 'infiltrated': 0.16666666666666666, '1949': 0.16666666666666666, 'armistice'

In [14]:
# Dividing the text into Sentences
sentence_tokens = [sent for sent in doc.sents]
print(sentence_tokens)

["Mahmoud Darwish is a Palestinian poet and member of the Palestinian National Council of the Palestine Liberation Committee., He has poetry collections full of modernist content., He was born in 1941 in the village of Al-Birwa, a Palestinian village located in the Galilee near the coast of Acre, where his family owned land., The family went out with the Palestinian refugees in 1948 to Lebanon, then returned infiltrated in 1949 after the armistice was signed, only to find the village destroyed and a moshav (an Israeli agricultural village) “Ahihud” built on its lands., Kibbutz Yas'ur lived with his family in the new village.
, After sending his secondary education in a high school in high school, he joined the Israeli Communist Party, the new party in the party, and the new one who edited it, edited it, edited it, edited Al-Fajr newspaper, which was published by Mapam.
, He served as president of the General Union of Palestinian Writers and Writers and edited Al-Karmel magazine., His r

# Calculating the Sentence Scores

In [15]:
# The sentence score for a particular sentence is the sum of the normalized frequencies of the words in that sentence. 
sentence_scores = {}
for sent in sentence_tokens:
    for word in sent:
        if word.text.lower() in word_frequencies.keys():
            if sent not in sentence_scores.keys():
                sentence_scores[sent] = word_frequencies[word.text.lower()]
            else:
                sentence_scores[sent] += word_frequencies[word.text.lower()]
                
sentence_scores

{"Mahmoud Darwish is a Palestinian poet and member of the Palestinian National Council of the Palestine Liberation Committee.: 0.6666666666666666,
 He has poetry collections full of modernist content.: 0.9999999999999999,
 He was born in 1941 in the village of Al-Birwa, a Palestinian village located in the Galilee near the coast of Acre, where his family owned land.: 3.3333333333333326,
 The family went out with the Palestinian refugees in 1948 to Lebanon, then returned infiltrated in 1949 after the armistice was signed, only to find the village destroyed and a moshav (an Israeli agricultural village) “Ahihud” built on its lands.: 4.833333333333334,
 Kibbutz Yas'ur lived with his family in the new village.: 2.0,
 After sending his secondary education in a high school in high school, he joined the Israeli Communist Party, the new party in the party, and the new one who edited it, edited it, edited it, edited Al-Fajr newspaper, which was published by Mapam.: 7.666666666666666,
 He served

In [16]:
from heapq import nlargest

In [17]:
# The length of the summary is 30% of the original length
select_length = int(len(sentence_tokens) * 0.3)
select_length

3

In [18]:
# Returning a list with the select_length largest elements 
summary = nlargest(select_length, sentence_scores, key = sentence_scores.get)
summary

[After sending his secondary education in a high school in high school, he joined the Israeli Communist Party, the new party in the party, and the new one who edited it, edited it, edited it, edited Al-Fajr newspaper, which was published by Mapam.,
 Critic Muhammad Fikri Al-Jazzar divided Darwish's poetry into three sections: The first stage, which is the stage of his presence in the homeland, which includes the beginnings of the poet's formation and awareness of the cause of his homeland and the formation of belonging to this homeland under the occupation.,
 Critics divided the stages of Darwish's poetry into several sections that combine the poet's relationship with his homeland, his cause "the Palestinian cause", exile and leaving home, all in light of his relationship with the self.]

In [19]:
# Combining these sentences to the final summary 
final_summary = [word.text for word in summary]
summary = ' '.join(final_summary)

# Displying the Original Text & The Generated Summary

In [20]:
# Displaying the original text 
print(text)

"Mahmoud Darwish is a Palestinian poet and member of the Palestinian National Council of the Palestine Liberation Committee. He has poetry collections full of modernist content. He was born in 1941 in the village of Al-Birwa, a Palestinian village located in the Galilee near the coast of Acre, where his family owned land. The family went out with the Palestinian refugees in 1948 to Lebanon, then returned infiltrated in 1949 after the armistice was signed, only to find the village destroyed and a moshav (an Israeli agricultural village) “Ahihud” built on its lands. Kibbutz Yas'ur lived with his family in the new village.
After sending his secondary education in a high school in high school, he joined the Israeli Communist Party, the new party in the party, and the new one who edited it, edited it, edited it, edited Al-Fajr newspaper, which was published by Mapam.
He served as president of the General Union of Palestinian Writers and Writers and edited Al-Karmel magazine. His residence w

In [21]:
# Displaying the generated summary
print(summary)

After sending his secondary education in a high school in high school, he joined the Israeli Communist Party, the new party in the party, and the new one who edited it, edited it, edited it, edited Al-Fajr newspaper, which was published by Mapam.
 Critic Muhammad Fikri Al-Jazzar divided Darwish's poetry into three sections: The first stage, which is the stage of his presence in the homeland, which includes the beginnings of the poet's formation and awareness of the cause of his homeland and the formation of belonging to this homeland under the occupation. Critics divided the stages of Darwish's poetry into several sections that combine the poet's relationship with his homeland, his cause "the Palestinian cause", exile and leaving home, all in light of his relationship with the self.


# Length of the Orginal Text 

In [22]:
len(text)

1790

# Length of the Generated Summary

In [23]:
len(summary)

793

The length has been reduced from 1790 to 793 after summarization